[Outlook] (clients) Add new Outlook on Windows references (#4446)
OfficeDev/office-js-docs-pr: docs/outlook/limits-for-activation-and-javascript-api-for-outlook-add-ins.md

|Internet headers|256 KB per message in Exchange Online<br/><br/>Header size limit determined by the organization's administrators in Exchange on-premises|[InternetHeaders.setAsync](/javascript/api/outlook/office.internetheaders) method|The total size limit of headers that can be applied to a message.|
|Extract well-known entities|2,000 number of characters|[Item.getEntities](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Item.getEntitiesByType](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Item.getFilteredEntitiesByName](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method|Limit for Exchange Server to extract well-known entities on the item body. Exchange Server ignores entities beyond that limit. Note that this limit is independent of whether the add-in uses an `ItemHasKnownEntity` rule.|

> |Exchange Web Services|5 MB number of characters on Outlook on the web, on Windows (starting in Version 2303 (Build 16225.10000)), on Mac (starting in Version 16.73 (23042601)), and on [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627)<br><br>1 MB of characters in earlier versions of Outlook on Windows and on Mac|[Mailbox.makeEwsRequestAsync](/javascript/api/outlook/office.mailbox#outlook-office-mailbox-makeewsrequestasync-member(1)) method|Limit for a request or response to a `Mailbox.makeEwsRequestAsync` call.|

Reviewer: ```suggestion
|Exchange Web Services|5 MB number of characters in Outlook on the web, on Windows (starting in Version 2303 (Build 16225.10000)), on Mac (starting in Version 16.73 (23042601)), and in [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627)<br><br>1 MB of characters in earlier versions of Outlook on Windows and on Mac|[Mailbox.makeEwsRequestAsync](/javascript/api/outlook/office.mailbox#outlook-office-mailbox-makeewsrequestasync-member(1)) method|Limit for a request or response to a `Mailbox.makeEwsRequestAsync` call.|
```

|Set the body|1 MB number of characters|[Body.prependAsync](/javascript/api/outlook/office.body#outlook-office-body-prependasync-member(1)) method<br/><br/>[Body.setAsync](/javascript/api/outlook/office.body#outlook-office-body-setasync-member(1))<br/><br/>[Body.setSelectedDataAsync](/javascript/api/outlook/office.body#outlook-office-body-setselecteddataasync-member(1)) method<br/><br/>[DisplayedBody.setAsync](/javascript/api/outlook/office.displayedbody?view=outlook-js-preview&preserve-view=true#outlook-office-displayedbody-setasync-member(1)) (preview)|Limit for setting the body of an appointment or message item.|
|Set the signature|30,000 characters|[Body.setSignatureAsync](/javascript/api/outlook/office.body#outlook-office-body-setsignatureasync-member(1)) method|Limit for the length of a signature in an appointment or message.|

> |Number of attachments|499 files on Outlook on the web, on mobile devices, and on [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627)|[Item.addFileAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method|Limit for the number of files that can be attached to an item for sending. Outlook on the web, on mobile devices, and on [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627) generally limit attaching up to 499 files through the user interface and `addFileAttachmentAsync` method. Outlook on Windows and on Mac don't specifically limit the number of file attachments. However, all Outlook clients observe the limit for the size of attachments determined by the user's Exchange Server (see the "Size of attachments" row in this table).|

Reviewer: ```suggestion
|Number of attachments|499 files in Outlook on the web, on mobile devices, and in [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627)|[Item.addFileAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method|Limit for the number of files that can be attached to an item for sending. Outlook on the web and on mobile devices, and [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627) generally limit attaching up to 499 files through the user interface and `addFileAttachmentAsync` method. Outlook on Windows and on Mac don't specifically limit the number of file attachments. However, all Outlook clients observe the limit for the size of attachments determined by the user's Exchange Server (see the "Size of attachments" row in this table).|
```

> |Size of attachments|Dependent on Exchange Server|[Item.addFileAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method|There is a limit on the size of all the attachments for a mail item, which an administrator can configure on the Exchange Server of the user's mailbox. For Outlook on Windows and on Mac, this limits the number of attachments for an item. For Outlook on the web, mobile devices, and [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627), the lesser of the two limits—the number of attachments and the size of all attachments—restricts the actual attachments for an item.|

Reviewer: ```suggestion
|Size of attachments|Dependent on Exchange Server|[Item.addFileAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method|There's a limit on the size of all the attachments for a mail item, which an administrator can configure on the Exchange Server of the user's mailbox. For Outlook on Windows and on Mac, this limits the number of attachments for an item. For Outlook on the web and on mobile devices, and [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627), the lesser of the two limits—the number of attachments and the size of all attachments—restricts the actual attachments for an item.|
```

|Attachment URI|2,048 characters|[Item.addFileAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Item.addFileAttachmentFromBase64Async](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method|Limit of the URI of the filename to be added as an attachment to an item.|
|Base64-encoded string of an attachment|27,892,122 characters (about 25 MB)|[Item.addFileAttachmentFromBase64Async](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method|Limit of the Base64-encoded string to be added as an attachment to an item.|
|Attachment ID|100 characters|[Item.addItemAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Item.removeAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method|Limit for the length of the ID of the attachment to be added to or removed from an item.|

> |Asynchronous calls|3 calls|[Item.addFileAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Item.addItemAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Item.removeAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Body.getTypeAsync](/javascript/api/outlook/office.body#outlook-office-body-gettypeasync-member(1)) method<br/><br/>[Body.prependAsync](/javascript/api/outlook/office.body#outlook-office-body-prependasync-member(1)) method<br/><br/>[Body.setSelectedDataAsync](/javascript/api/outlook/office.body#outlook-office-body-setselecteddataasync-member(1)) method<br/><br/>[CustomProperties.saveAsync](/javascript/api/outlook/office.customproperties#outlook-office-customproperties-saveasync-member(1)) method<br/><br/>[Item.LoadCustomPropertiesAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Location.getAsync](/javascript/api/outlook/office.location#outlook-office-location-getasync-member(1)) method<br/><br/>[Location.setAsync](/javascript/api/outlook/office.location#outlook-office-location-setasync-member(1)) method<br/><br/>[Mailbox.getCallbackTokenAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox#methods) method<br/><br/>[Mailbox.getUserIdentityTokenAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox#methods) method<br/><br/>[Mailbox.makeEwsRequestAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox#methods) method<br/><br/>[Recipients.addAsync](/javascript/api/outlook/office.recipients#outlook-office-recipients-addasync-member(1)) method<br/><br/>[Recipients.getAsync](/javascript/api/outlook/office.recipients#outlook-office-recipients-getasync-member(1)) method<br/><br/>[Recipients.setAsync](/javascript/api/outlook/office.recipients#outlook-office-recipients-setasync-member(1)) method<br/><br/>[RoamingSettings.saveAsync](/javascript/api/outlook/office.roamingsettings#outlook-office-roamingsettings-saveasync-member(1)) method<br/><br/>[Subject.getAsync](/javascript/api/outlook/office.subject#outlook-office-subject-getasync-member(1)) method<br/><br/>[Subject.setAsync](/javascript/api/outlook/office.subject#outlook-office-subject-setasync-member(1)) method<br/><br/>[Time.getAsync](/javascript/api/outlook/office.time#outlook-office-time-getasync-member(1)) method<br/><br/>[Time.setAsync](/javascript/api/outlook/office.time#outlook-office-time-setasync-member(1)) method|For Outlook on the web, mobile devices, and [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627): limit of the number of simultaneous asynchronous calls at any one time, as browsers allow only a limited number of asynchronous calls to servers. |

Reviewer: ```suggestion
|Asynchronous calls|3 calls|[Item.addFileAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Item.addItemAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Item.removeAttachmentAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Body.getTypeAsync](/javascript/api/outlook/office.body#outlook-office-body-gettypeasync-member(1)) method<br/><br/>[Body.prependAsync](/javascript/api/outlook/office.body#outlook-office-body-prependasync-member(1)) method<br/><br/>[Body.setSelectedDataAsync](/javascript/api/outlook/office.body#outlook-office-body-setselecteddataasync-member(1)) method<br/><br/>[CustomProperties.saveAsync](/javascript/api/outlook/office.customproperties#outlook-office-customproperties-saveasync-member(1)) method<br/><br/>[Item.LoadCustomPropertiesAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox.item#methods) method<br/><br/>[Location.getAsync](/javascript/api/outlook/office.location#outlook-office-location-getasync-member(1)) method<br/><br/>[Location.setAsync](/javascript/api/outlook/office.location#outlook-office-location-setasync-member(1)) method<br/><br/>[Mailbox.getCallbackTokenAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox#methods) method<br/><br/>[Mailbox.getUserIdentityTokenAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox#methods) method<br/><br/>[Mailbox.makeEwsRequestAsync](/javascript/api/requirement-sets/outlook/preview-requirement-set/office.context.mailbox#methods) method<br/><br/>[Recipients.addAsync](/javascript/api/outlook/office.recipients#outlook-office-recipients-addasync-member(1)) method<br/><br/>[Recipients.getAsync](/javascript/api/outlook/office.recipients#outlook-office-recipients-getasync-member(1)) method<br/><br/>[Recipients.setAsync](/javascript/api/outlook/office.recipients#outlook-office-recipients-setasync-member(1)) method<br/><br/>[RoamingSettings.saveAsync](/javascript/api/outlook/office.roamingsettings#outlook-office-roamingsettings-saveasync-member(1)) method<br/><br/>[Subject.getAsync](/javascript/api/outlook/office.subject#outlook-office-subject-getasync-member(1)) method<br/><br/>[Subject.setAsync](/javascript/api/outlook/office.subject#outlook-office-subject-setasync-member(1)) method<br/><br/>[Time.getAsync](/javascript/api/outlook/office.time#outlook-office-time-getasync-member(1)) method<br/><br/>[Time.setAsync](/javascript/api/outlook/office.time#outlook-office-time-setasync-member(1)) method|For Outlook on the web and on mobile devices, and [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627): limit of the number of simultaneous asynchronous calls at any one time, as browsers allow only a limited number of asynchronous calls to servers. |
```

The following table lists the limits and describes the differences in the support for regular expressions across Outlook clients. The support is independent of any specific type of device and item body.

> |Outlook on Windows and on Mac|Outlook on the web, on mobile devices, and on new Windows client (preview)|

Reviewer: ```suggestion
|Outlook on Windows and on Mac|Outlook on the web, on mobile devices, and on new Windows client (preview)|
```

|:-----|:-----|
|Clients use the C++ regular expression engine provided as part of the Visual Studio standard template library. This engine complies with ECMAScript 5 standards. |Clients use regular expression evaluation that's part of JavaScript. It's provided by the browser and supports a superset of ECMAScript 5.|

> |Because of the different regex engines, a regex that includes a custom character class based on predefined character classes may return different results in Outlook on Windows and on Mac versus Outlook on the web, on mobile devices, and on [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627).<br/><br/>As an example, the regex `[\s\S]{0,100}` matches any number, between 0 and 100, of single characters that is a whitespace or a non-whitespace. This regex returns different results depending on your Outlook client.<br/><br/>As a workaround, you should rewrite the regex as `(\s\|\S){0,100}`. This regex matches any number, between 0 and 100, of white space or non-white space.<br/><br/>You should test each regex thoroughly on each Outlook client. If a regex returns different results, rewrite the regex. |You should test each regex thoroughly on each Outlook client. If a regex returns different results, rewrite the regex.|

Reviewer: ```suggestion
|Because of the different regex engines, a regex that includes a custom character class based on predefined character classes may return different results in Outlook on Windows and on Mac versus in Outlook on the web, on mobile devices, and in [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627).<br/><br/>As an example, the regex `[\s\S]{0,100}` matches any number, between 0 and 100, of single characters that is a whitespace or a non-whitespace. This regex returns different results depending on your Outlook client.<br/><br/>As a workaround, you should rewrite the regex as `(\s\|\S){0,100}`. This regex matches any number, between 0 and 100, of white space or non-white space.<br/><br/>You should test each regex thoroughly on each Outlook client. If a regex returns different results, rewrite the regex. |You should test each regex thoroughly on each Outlook client. If a regex returns different results, rewrite the regex.|
```

> |By default, the evaluation of all regular expressions for an add-in is limited to one second. Exceeding this limit causes reevaluation of up to three times. Beyond the reevaluation limit, Outlook on Windows and on Mac prevent the add-in from running for the same mailbox in any of the Outlook clients.<br/><br/>Administrators can override these evaluation limits by using the `OutlookActivationAlertThreshold` and `OutlookActivationManagerRetryLimit` registry keys.|The resource monitoring and registry settings supported in Outlook on Windows and on Mac aren't supported in Outlook on the web, on mobile devices, and on new Outlook on Windows (preview). However, add-ins with regular expressions that require excessive amount of evaluation time on Outlook on Windows and on Mac are disabled for the same mailbox on all the Outlook clients.|

Reviewer: ```suggestion
|By default, the evaluation of all regular expressions for an add-in is limited to one second. Exceeding this limit causes reevaluation of up to three times. Beyond the reevaluation limit, Outlook on Windows and on Mac prevent the add-in from running for the same mailbox in any of the Outlook clients.<br/><br/>Administrators can override these evaluation limits by using the `OutlookActivationAlertThreshold` and `OutlookActivationManagerRetryLimit` registry keys.|The resource monitoring and registry settings supported in Outlook on Windows and on Mac aren't supported in Outlook on the web, on mobile devices, and in new Outlook on Windows (preview). However, add-ins with regular expressions that require excessive amount of evaluation time in Outlook on Windows and on Mac are disabled for the same mailbox on all the Outlook clients.|
```

### Limits on the size of the item body evaluated

The following table lists the limits and describes the differences in the portion of the item body to which each Outlook client applies a regular expression. Some of these limits depend on the type of device and item body, if the regular expression is applied on the item body.

> ||Outlook on Windows and on Mac|Outlook on mobile devices|Outlook on the web and on new Windows client (preview)|

Reviewer: ```suggestion
||Outlook on Windows and on Mac|Outlook on mobile devices|Outlook on the web and on new Windows client (preview)|
```

The following table lists the limits and describes the differences in the matches that each Outlook client returns after evaluating a regular expression. The support is independent of any specific type of device, but may depend on the type of item body, if the regular expression is applied on the item body.

||Outlook on Windows and on Mac|Outlook on the web, on mobile devices, and on new Windows client (preview)|
|:-----|:-----|:-----|

> |**Order of returned matches**|Assume the `getRegExMatches` method returns different matches for the same regular expression applied on the same mail item in Outlook on Windows and on Mac versus in Outlook on the web, on mobile devices, and on [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627).|Assume the `getRegExMatches` method returns different matches for the same regular expression applied on the same mail item in Outlook on Windows and on Mac versus in Outlook on the web, on mobile devices, and on new Outlook on Windows (preview).|
> |**Plain text item body**|The `getRegExMatches` method returns any matches that are up to 1,536 (1.5 KB) characters, for a maximum of 50 matches.<br/><br/>**Note**: `getRegExMatches` doesn't return matches in any specific order in the returned array. In general, assume the order of matches for the same regular expression applied on the same mail item in Outlook on Windows and on Mac is different from that in Outlook on the web, on mobile devices, and on new Outlook on Windows (preview).|`getRegExMatches` returns any matches that are up to 3,072 (3 KB) characters, for a maximum of 50 matches.|
> |**HTML item body**|`getRegExMatches` returns any matches that are up to 3,072 (3 KB) characters, for a maximum of 50 matches.<br/> <br/> **Note**: `getRegExMatches` doesn't return matches in any specific order in the returned array. In general, assume the order of matches for the same regular expression applied on the same mail item in Outlook on Windows and on Mac is different from that in Outlook on the web, on mobile devices, and on new Outlook on Windows (preview).|`getRegExMatches` returns any matches that are up to 3,072 (3 KB) characters, for a maximum of 50 matches.|

Reviewer: ```suggestion
|**Order of returned matches**|Assume the `getRegExMatches` method returns different matches for the same regular expression applied on the same mail item in Outlook on Windows and on Mac versus in Outlook on the web, on mobile devices, and in [new Outlook on Windows (preview)](https://support.microsoft.com/office/656bb8d9-5a60-49b2-a98b-ba7822bc7627).|Assume the `getRegExMatches` method returns different matches for the same regular expression applied on the same mail item in Outlook on Windows and on Mac versus in Outlook on the web, on mobile devices, and in new Outlook on Windows (preview).|
|**Plain text item body**|The `getRegExMatches` method returns any matches that are up to 1,536 (1.5 KB) characters, for a maximum of 50 matches.<br/><br/>**Note**: `getRegExMatches` doesn't return matches in any specific order in the returned array. In general, assume the order of matches for the same regular expression applied on the same mail item in Outlook on Windows and on Mac is different from that in Outlook on the web, on mobile devices, and in new Outlook on Windows (preview).|`getRegExMatches` returns any matches that are up to 3,072 (3 KB) characters, for a maximum of 50 matches.|
|**HTML item body**|`getRegExMatches` returns any matches that are up to 3,072 (3 KB) characters, for a maximum of 50 matches.<br/> <br/> **Note**: `getRegExMatches` doesn't return matches in any specific order in the returned array. In general, assume the order of matches for the same regular expression applied on the same mail item in Outlook on Windows and on Mac is different from that in Outlook on the web, on mobile devices, and in new Outlook on Windows (preview).|`getRegExMatches` returns any matches that are up to 3,072 (3 KB) characters, for a maximum of 50 matches.|
```
